File: core/src/main/java/com/lomekwi/cave/ui/tabs/TopTabbedPane.java
package com.lomekwi.cave.ui.tabs;

import com.google.common.eventbus.Subscribe;
import com.kotcrab.vis.ui.widget.tabbedpane.Tab;
import com.kotcrab.vis.ui.widget.tabbedpane.TabbedPane;
import com.kotcrab.vis.ui.widget.tabbedpane.TabbedPaneListener;
import com.lomekwi.cave.project.ProjectBackgroundedEvent;
import com.lomekwi.cave.project.ProjectFrontedEvent;
import com.lomekwi.cave.project.ProjectLoadedEvent;
import com.lomekwi.cave.ui.Root;
import com.lomekwi.cave.app.App;

public class TopTabbedPane extends TabbedPane {
    private ProjectTab currentProjectTab;

    public TopTabbedPane() {
        super();
        App.appEventBus.register(this);
        addListener(new TabbedPaneListener() {
            @Override
            public void switchedTab(Tab tab) {
                Root.getInstance().getMajorArea().clear();
                Root.getInstance().getMajorArea().add(tab.getContentTable()).grow();

                if (currentProjectTab != null && currentProjectTab != tab) {
                    currentProjectTab.getProject().projEventBus.post(ProjectBackgroundedEvent.INSTANCE);
                }

                if (tab instanceof ProjectTab) {
                    currentProjectTab = (ProjectTab) tab;
                    ((ProjectTab) tab).getProject().projEventBus.post(ProjectFrontedEvent.INSTANCE);
                } else {
                    currentProjectTab = null;
                }

                App.appEventBus.post(TabSwitchedEvent.INSTANCE);
            }

            @Override
            public void removedTab(Tab tab) {
                if (tab instanceof ProjectTab) {
                    ((ProjectTab) tab).getProject().projEventBus.post(ProjectBackgroundedEvent.INSTANCE);
                    if (currentProjectTab == tab) {
                        currentProjectTab = null;
                    }
                }
            }

            @Override
            public void removedAllTabs() {
                Root.getInstance().getMajorArea().clear();
                App.appEventBus.post(TabSwitchedEvent.INSTANCE);
            }
        });
    }

    public void add(Tab... tabs){
        for(Tab tab : tabs){
            super.add(tab);
        }
    }
    @Subscribe
    public void onNewProject(ProjectLoadedEvent event){
        ProjectTab pt = new ProjectTab(event.getNewProject());
        add(pt);
        switchTab(pt);
    }
    @Subscribe
    public void onSettingsOpened(SettingsOpenedEvent event){
        SettingsTab st = new SettingsTab();
        add(st);
        switchTab(st);
    }
}

File: core/src/main/scala/com/lomekwi/cave/ui/tabs/app/TopTabbedPane.scala
package com.lomekwi.cave.ui.tabs.app

import com.google.common.eventbus.Subscribe
import com.kotcrab.vis.ui.widget.tabbedpane.Tab
import com.kotcrab.vis.ui.widget.tabbedpane.TabbedPaneListener
import com.lomekwi.cave.project.ProjectBackgroundedEvent
import com.lomekwi.cave.project.ProjectFrontedEvent
import com.lomekwi.cave.project.ProjectLoadedEvent
import com.lomekwi.cave.app.App
import com.lomekwi.cave.ui.widget.AutoHideTabbedPane
import scala.compiletime.uninitialized

class TopTabbedPane extends AutoHideTabbedPane {
  private var currentProjectTab: ProjectTab = uninitialized

  {
    App.appEventBus.register(this)
    addListener(new TabbedPaneListener {
      override def switchedTab(tab: Tab): Unit = {
        if (currentProjectTab != null && (currentProjectTab ne tab)) {
          currentProjectTab.getProject.playhead.setPlaying(false)
        }

        App.root.getMajorArea.setActor(tab.getContentTable)

        if (currentProjectTab != null && (currentProjectTab ne tab)) {
          currentProjectTab.getProject.projEventBus.post(ProjectBackgroundedEvent)
        }

        tab match {
          case pt: ProjectTab =>
            currentProjectTab = pt
            pt.getProject.projEventBus.post(ProjectFrontedEvent)
          case _ =>
            currentProjectTab = null
        }

        App.appEventBus.post(TabSwitchedEvent)
      }

      override def removedTab(tab: Tab): Unit = {
        tab match {
          case pt: ProjectTab =>
            pt.getProject.playhead.setPlaying(false)
            pt.getProject.projEventBus.post(ProjectBackgroundedEvent)
            if (currentProjectTab eq tab) {
              currentProjectTab = null
            }
          case _ =>
        }
      }

      override def removedAllTabs(): Unit = {
        App.root.getMajorArea.setActor(null)
        App.appEventBus.post(TabSwitchedEvent)
      }
    })
  }

  def add(tabs: Tab*): Unit = {
    for (tab <- tabs) {
      super.add(tab)
    }
  }
  @Subscribe
  def onNewProject(event: ProjectLoadedEvent): Unit = {
    val pt = new ProjectTab(event.newProject)
    super.add(pt)
    switchTab(pt)
  }
}
